File: 2025/373. Find K Pairs with Smallest Sums/Solution.java
class SumPair {
    public int i;
    public int j;
    public int sum; 
    public SumPair(int i, int j, int sum) {
        this.i = i;
        this.j = j;
        this.sum = sum;
    }
    @Override
    public int hashCode() {
        return i * 31 ^ j;
    }
    @Override
    public boolean equals(Object o) {
        if (o instanceof SumPair) {
            SumPair other = (SumPair) o;
            return (i == other.i && j == other.j);
        }
        return false;
    }
}

class Solution {
    public List<List<Integer>> kSmallestPairs(int[] nums1, int[] nums2, int k) {
        List<List<Integer>> res = new ArrayList<>();
        Set<SumPair> set = new HashSet<>();
        Queue<SumPair> heap = new PriorityQueue<>((x, y) -> (x.sum - y.sum));
        heap.add(new SumPair(0, 0, nums1[0] + nums2[0]));
        while (!heap.isEmpty() && k > 0) {
            k--;
            SumPair sp = heap.remove();
            int i = sp.i, j = sp.j;
            res.add(Arrays.asList(new Integer[]{nums1[i], nums2[j]}));
            if (i < nums1.length - 1) {
                SumPair nsp = new SumPair(i + 1, j, nums1[i + 1] + nums2[j]);
                if (!set.contains(nsp)) {
                    heap.add(nsp);
                    set.add(nsp);
                }
            }
            if (j < nums2.length - 1) {
                SumPair nsp = new SumPair(i, j + 1, nums1[i] + nums2[j + 1]);
                if (!set.contains(nsp)) {
                    heap.add(nsp);
                    set.add(nsp);
                }
            }
        }
        return res;
    }
}

File: 2025/373. Find K Pairs with Smallest Sums/solution.py
class Solution:
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        res, heap = [], []
        if not nums1 or not nums2:
            return res
        for i in range(min(k, len(nums1))):
            heapq.heappush(heap, (nums1[i] + nums2[0], i, 0))
        while heap and k > 0:
            (_, i, j) = heapq.heappop(heap)
            res.append((nums1[i], nums2[j]))
            k -= 1
            if j < len(nums2) - 1:
                heapq.heappush(heap, (nums1[i] + nums2[j + 1], i, j + 1))
        return res